syntax = "proto3";

package TrackerServer;

service TrackerServer {
    rpc AskFileOperation(AskFileOperationRequest) returns (AskFileOperationResponse) {}
    rpc GetFileList(GetFileListRequest) returns (GetFileListResponse) {}
    rpc UpdateFileInfo(UpdateFileInfoRequest) returns (UpdateFileInfoResponse) {}
    rpc Replicate(ReplicateRequest) returns (ReplicateResponse) {}
    rpc AddStorageServer(AddStorageServerRequest) returns (AddStorageServerResponse) {}
    rpc DeleteFile(DeleteFileRequest) returns (DeleteFileResponse) {}
}

message AskFileOperationRequest {
    string operation = 1; 
    string filename = 2;
    int32 group_id = 3;
    string path = 4;
}

message AskFileOperationResponse {
    bool status = 1;
    string ip = 2;
    int32 port = 3;
    string time = 4;
}

message GetFileListRequest {
    string operation = 1; // null
}

message File {
    string file_name = 1;
    string file_path = 2;
}

message Group {
    int32 group_id = 1;
    repeated File files = 2;
}

message GetFileListResponse {
    repeated Group groups = 1;
}

message UpdateFileInfoRequest {
    string filename = 1;
    int32 group_id = 2;
    string filepath = 3;
}

message UpdateFileInfoResponse {
    bool status = 1;
    string time = 2;
}

message ReplicateRequest {
    int32 port = 1;
    string ip = 2;
}

message ReplicateResponse {
    // return the message with same group
    bool status = 1;
    int32 port = 2;
    string ip = 3;
}

message AddStorageServerRequest {
    int32 group_id = 1;
    string ip = 2;
    int32 port = 3;
    int32 storage_server_id = 4;
}

message AddStorageServerResponse {
    bool status = 1;
}

message DeleteFileRequest {
    string filename = 1;
    string filepath = 2;
    int32 group_id = 3;
}

message DeleteFileResponse {
    bool status = 1;
}